pkg/syncer/storagepool/intended_state.go

```diff
@@ -211,23 +217,44 @@ func newIntendedVsanSNAState(ctx context.Context, scWatchCntlr *StorageClassWatc
 			compatSC = append(compatSC, scName)
 		}
 	}
+	// In vSAN-SNA overhead is upperbound by roughly 4% of volume requested capacity. Thus maximum allocatable space in vSAN-SNA is 96% of free-capacity.
+	vsanSnaDsType := fmt.Sprintf("%s-sna", vsan.dsType)
+	freeSpace := resource.NewQuantity(vsanHost.Capacity - vsanHost.CapacityUsed, resource.DecimalSI)
+	allocatableSpace := getAllocatableSpace(freeSpace, vsanSnaDsType)
 
 	return &intendedState{
-		dsMoid:        fmt.Sprintf("%s-%s", vsan.dsMoid, node),
-		dsType:        fmt.Sprintf("%s-sna", vsan.dsType),
-		spName:        fmt.Sprintf("%s-%s", vsan.spName, node),
-		capacity:      resource.NewQuantity(vsanHost.Capacity, resource.DecimalSI),
-		freeSpace:     resource.NewQuantity(vsanHost.Capacity-vsanHost.CapacityUsed, resource.DecimalSI),
-		url:           vsan.url,
-		accessible:    true,  // If this node is in accessible node list of the vsan-ds, then sp is accessible
-		datastoreInMM: false, // If this node is inMM - the storagepool will not exist at all
-		allHostsInMM:  false, // If this node is inMM - the storagepool will not exist at all
-		nodes:         nodes,
-		compatSC:      compatSC,
-		isRemoteVsan:  false,
+		dsMoid:           fmt.Sprintf("%s-%s", vsan.dsMoid, node),
+		dsType:           vsanSnaDsType,
+		spName:           fmt.Sprintf("%s-%s", vsan.spName, node),
+		capacity:         resource.NewQuantity(vsanHost.Capacity, resource.DecimalSI),
+		freeSpace:        freeSpace,
+		allocatableSpace: allocatableSpace,
+		url:              vsan.url,
+		accessible:       true,  // If this node is in accessible node list of the vsan-ds, then sp is accessible
+		datastoreInMM:    false, // If this node is inMM - the storagepool will not exist at all
+		allHostsInMM:     false, // If this node is inMM - the storagepool will not exist at all
+		nodes:            nodes,
+		compatSC:         compatSC,
+		isRemoteVsan:     false,
 	}, nil
 }
 
+func getAllocatableSpace(freeSpace *resource.Quantity, dsType string) *resource.Quantity {
+	if dsType == spTypePrefix+"vsanD" {
+		// In vSAN Direct overhead to provision volume is upperbound by 4MiB space. Thus maximum allocatable space in vSAN Direct is 4MiB less than free-capacity.
```

Comment:
Where do the values 4mb and 4% come from? Can we add the source of information as a comment?

Also, can we please save these values as constants. If we need to update these in the future, it is at a single place.

Reply:
These numbers are empirically obtained by refreshing the capacity information in SP and provisioning volumes of different sizes on vsan-sna and vsan direct. Some of the observations are listed below.

```
# vsan-sna capacity before 20GiB provisioning
    accessibleNodes:
    - wdc-10-206-98-54.eng.vmware.com
    capacity:
      freeSpace: 141100748638
      total: 214731587584
# after 20GiB volume provisioning
    accessibleNodes:
    - wdc-10-206-98-54.eng.vmware.com
    capacity:
      freeSpace: 119172927326
      total: 214731587584
Overhead in MiB = (141100748638 - 119172927326)/(1024*1024) - 20*1024 = 432MiB
```
```
# vsan-sna capacity before 40GiB provisioning
    accessibleNodes:
    - wdc-10-206-111-171.eng.vmware.com
    capacity:
      freeSpace: 139674685278
      total: 214731587584
# after 40GiB volume provisioning
    accessibleNodes:
    - wdc-10-206-111-171.eng.vmware.com
    capacity:
      freeSpace: 95844208478
      total: 214731587584
Overhead in MiB = (139674685278- 95844208478)/(1024*1024) - 40*1024 = 840MiB
```

```
# vsan-direct disk capacity before 40GiB provisioning
    accessibleNodes:
    - wdc-10-206-100-35.eng.vmware.com
    capacity:
      freeSpace: 40901804032
      total: 85630910464
# after 40GiB volume provisioning
    accessibleNodes:
    - wdc-10-206-100-35.eng.vmware.com
    capacity:
      freeSpace: 83852525568
      total: 85630910464
Overhead in MiB = (83852525568- 40901804032)/(1024*1024) - 40*1024 = 1MiB
```
I have saved these values as constant and added the source of info as comments.